reference/src/representation/enums.md

```diff
@@ -0,0 +1,342 @@
+etc) -- if no type is specified, then the compiler will select one
+based on what a size a fieldless enum would have with the same number of
+variants.
+
+This layout, while more compatible and arguably more obvious, is also
+less efficient than the non-C compatible layout in some cases in terms
+of total size. For example, the `TwoCases` example given in the
+preivous section only occupies 4 bytes with `#[repr(u8)]`, but would
+occupy 6 bytes with `#[repr(C, u8)]`, as more padding is required.
+
+**Example.** The following enum:
+
+```rust
+#[repr(C, Int)]
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+}
+```
+
+is equivalent to the following Rust definition:
+
+```rust
+#[repr(C)]
+struct MyEnumRepr {
+    tag: MyEnumTag,
+    payload: MyEnumPayload,
+}
+
+#[repr(Int)]
+enum MyEnumTag { A, B, C, D }
+
+#[repr(C)]
+union MyEnumPayload {
+   A: u32,
+   B: MyEnumPayloadB,
+   C: MyEnumPayloadC,
+   D: (),
+}
+
+#[repr(C)]
+struct MyEnumPayloadB(f32, u64);
+
+#[repr(C)]
+struct MyEnumPayloadC { x: u32, y: u8 }
+}
+```
+
+This enum can also be represented in C++ as follows:
+
+```c++
+#include <stdint.h>
+
+enum class MyEnumTag: CppEquivalentOfInt { A, B, C, D };
+struct MyEnumPayloadB { float _0; uint64_t _1;  };
+struct MyEnumPayloadC { uint32_t x; uint8_t y; };
+
+union MyEnumPayload {
+   uint32_t A;
+   MyEnumPayloadB B;
+   MyEnumPayloadC C;
+};
+
+struct MyEnum {
+    MyEnumTag tag;
+    MyEnumPayload payload;
+};
+```
+
+### Layout of a data-carrying enums without a repr annotation
+
+If no explicit `#[repr]` attribute is used, then the layout of a
+data-carrying enum is typically **not specified**. However, in certain
+select cases, there are **guaranteed layout optimizations** that may
+apply, as described below.
+
+#### Discriminant elision on Option-like enums
+
+(Meta-note: The content in this section is not described by any RFC
+and is therefore "non-normative".)
+
+**Definition.** An **option-like enum** is a 2-variant enum where:
+
+- one variant has a single field, and
+- the other variant has no fields (the "unit variant").
+
+The simplest example is `Option<T>` itself, where the `Some` variant
+has a single field (of type `T`), and the `None` variant has no
+fields. But other enums that fit that same template fit.
+
+**Definition.** The **payload** of an option-like enum is the single
+field which it contains; in the case of `Option<T>`, the payload has
+type `T`.
+
+**Definition.** In some cases, the payload type may contain illegal
+values, which are called **niches**. For example, a value of type `&T`
+may never be NULL, and hence defines a niche consisting of the
+bitstring `0`.  Similarly, the standard library types [`NonZeroU8`]
+and friends may never be zero, and hence also define the value of `0`
+as a niche. (Types that define niche values will say so as part of the
+description of their representation invariant, which -- as of this
```

Comment:
"representation invariant"? I don't think I've seen that term before.

Are you suggestion we rename "validity invariant" to "representation invariant"? (Not that I am necessarily opposed, just wondering where this term is coming from.)

Reply:
I meant validity invariant